File: os/cv.cpp
#include <pthread.h>
#include <unistd.h>

#include <iostream>

// 뮤텍스와 조건 변수 선언
pthread_mutex_t mutex;
pthread_cond_t cond;
bool ready = false;

void* thread_job1(void* arg) {
    std::cout << "P1: 먼저 시작" << std::endl;

    pthread_mutex_lock(&mutex);
    std::cout << "P1: 2초 대기" << std::endl;
    while (!ready) {
        pthread_cond_wait(&cond, &mutex);  // 조건 변수 wait
    }
    pthread_mutex_unlock(&mutex);

    std::cout << "P1: 다시 시작" << std::endl;
    std::cout << "P1: 종료" << std::endl;

    return NULL;
}

void* thread_job2(void* arg) {
    std::cout << "P2: 2초 실행 시작" << std::endl;
    sleep(2);  // 2초 대기

    std::cout << "P2: 실행 완료" << std::endl;
    pthread_mutex_lock(&mutex);
    ready = true;
    pthread_cond_signal(&cond);  // 조건 변수 signal
    pthread_mutex_unlock(&mutex);
    return NULL;
}

int main() {
    pthread_t t1, t2;

    // 뮤텍스와 조건 변수 초기화
    pthread_mutex_init(&mutex, NULL);
    pthread_cond_init(&cond, NULL);

    // 스레드 생성
    pthread_create(&t1, NULL, thread_job1, NULL);
    pthread_create(&t2, NULL, thread_job2, NULL);

    // 스레드 종료 대기
    pthread_join(t1, NULL);
    pthread_join(t2, NULL);

    // 뮤텍스와 조건 변수 해제
    pthread_mutex_destroy(&mutex);
    pthread_cond_destroy(&cond);

    return 0;
}

File: os/CV.java
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class CV {
    private static final Lock lock = new ReentrantLock();
    private static final Condition cond = lock.newCondition();
    private static boolean ready = false;

    public static void main(String[] args) throws InterruptedException {
        Thread t1 = new Thread(new ThreadJob1());
        Thread t2 = new Thread(new ThreadJob2());
        t1.start();
        t2.start();
        t1.join();
        t2.join();
    }

    static class ThreadJob1 implements Runnable {
        @Override
        public void run() {
            System.out.println("P1: 먼저 시작");
            lock.lock();
            try {
                System.out.println("P1: 2초 대기");
                while (!ready) {
                    cond.await(); // 조건 변수 wait
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            } finally {
                lock.unlock();
            }
            System.out.println("P1: 다시 시작");
            System.out.println("P1: 종료");
        }
    }

    static class ThreadJob2 implements Runnable {
        @Override
        public void run() {
            System.out.println("P2: 2초 실행 시작");
            try {
                Thread.sleep(2000); // 2초 대기
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            System.out.println("P2: 실행 완료");
            lock.lock();
            try {
                ready = true;
                cond.signal(); // 조건 변수 signal
            } finally {
                lock.unlock();
            }
        }
    }
}
